### PR TITLE
python3: upgrade to 3.8

### PR DESCRIPTION
**Pull request by iMichka**
- [x] Have you followed the [guidelines for contributing](https://github.com/Homebrew/homebrew-core/blob/master/CONTRIBUTING.md)?
- [x] Have you checked that there aren't other open [pull requests](https://github.com/Homebrew/homebrew-core/pulls) for the same formula update/change?
- [ ] Have you built your formula locally with `brew install --build-from-source <formula>`, where `<formula>` is the name of the formula you're submitting?
- [ ] Is your test running fine `brew test <formula>`, where `<formula>` is the name of the formula you're submitting?
- [ ] Does your build pass `brew audit --strict <formula>` (after doing `brew install <formula>`)?

-----

### CODE REVIEW

**Comment by iMichka** [comment]
#47274 needs to be finished first.

File: Aliases/python3 [hunk 1]
```diff
@@ -1 +1 @@
-../Formula/python.rb
\ No newline at end of file
+Formula/python@3.8.rb
```
Comment (bayandin) [inline Aliases/python3:1]:
```suggestion
../Formula/python@3.8.rb
```

File: Aliases/python@3 [hunk 2]
```diff
@@ -1 +1 @@
-../Formula/python.rb
\ No newline at end of file
+Formula/python@3.8.rb
```
Comment (bayandin) [inline Aliases/python@3:1]:
```suggestion
../Formula/python@3.8.rb
```

**Review by bayandin — COMMENTED** [review]
It looks like `aurora-cli`, `gradio`, and `protobuf@3.7` require a revision bump as well

**Comment by Bo98** [comment]
`aurora-cli` can't be built because the source is missing and there's a brew bug preventing the mirror from working.

`gradio` indeed should have a revision bump. Deprecation doesn't mean it can't be installed.

**Comment by bayandin** [comment]
Ah, I've missed that these formulae are disabled, thanks for explanation @Bo98!

Actually it's still a bit unclear for me if we need to update disabled/deprecated formulae.

**Comment by Bo98** [comment]
We should still treat deprecated formulae like undeprecated formulae in terms of updating. Deprecated formulae are still available for install and use and as such should not be intentionally broken if we are aware of an issue (the difference being that we don't actively test it as a dependent).

Disabled formulae should probably still change from `python` to `python@3.7` for style reasons, but no one will actually benefit from the update.

**Comment by iMichka** [comment]
```
2020-05-30T16:28:38.2144240Z [31m==>[0m [1m[31mFAILED[0m[0m
2020-05-30T16:28:38.2147610Z [31mError:[0m 2 problems in 1 formula detected
2020-05-30T16:28:38.2148060Z python@3.7:
2020-05-30T16:28:38.2148470Z   * python@3.7 is versioned but no python formula exists
2020-05-30T16:28:38.2148800Z   * New formulae in homebrew/core should not have a `bottle do` block

2020-05-30T16:38:14.0911640Z python@3.8:
2020-05-30T16:38:14.0912030Z   * Versioned formulae in homebrew/core should use `keg_only :versioned_formula`
2020-05-30T16:38:14.0912380Z   * Formulae should not have a `HEAD` spec
2020-05-30T16:38:14.0912730Z   * python@3.8 is versioned but no python formula exists
```

**Comment by Bo98** [comment]
> `* python@3.7 is versioned but no python formula exists`
> `* python@3.8 is versioned but no python formula exists`

Add `Aliases/python`.

> `* New formulae in homebrew/core should not have a bottle do block`

Removing the bottle block for python@3.7 is the easiest solution here. We cannot use existing bottles anyway.

> `* Versioned formulae in homebrew/core should use keg_only :versioned_formula`

Requires a whitelist: Homebrew/brew#7670.

> * Formulae should not have a `HEAD` spec

I don't think we should have HEAD support. `--HEAD` will install some 3.9 dev version rather than 3.8, which will break most dependents. The formula is also named `python@3.8`.

**Comment by iMichka** [comment]
I made the needed changes.

Just a question about the upgrade for our uses:
- I installed with `brew install python3` back then and I have python 3.7 installed
- I now run `brew update`
- `brew outdated` shows that nothing has changed

At this point python 3.7 is still installed and in my path. At which point do we upgrade the existing python 3.7 installation to python 3.8 for our users; switching the paths from keg only to non keg only and the other way round?

**Comment by Bo98** [comment]
@MikeMcQuaid might know. He indicated that `brew` should remember you installed via an alias and upgrade based on that.

**Comment by MikeMcQuaid** [comment]
The `source: path` in the `INSTALL_RECEIPT.json` is used to calculate whether it was installed from an alias or not. If it was installed from the alias: that's what we check to determine if it should be upgraded or not. `brew install python@3.8` and `brew install python` will behave differently for upgrades in future, as a result, despite them pointing to the same formula file.

**Comment by MikeMcQuaid** [comment]
See https://github.com/Homebrew/brew/blob/4c49224193b88981e8c4a93db4398eb793536057/Library/Homebrew/formula.rb#L271-L313 and code that calls these methods.

**Comment by iMichka** [comment]
I still need to update the caveats, they are wrong.

And I'll test the upgrade path, I am unsure of what is going on.

**Comment by iMichka** [comment]
This is what happens if we merge this PR:
- Someone ran `brew install python` before this was merged. He has 3.7 in /usr/local
- He will probably have the 3.8 keg only version too

When running `brew upgrade`; the 3.8 symlinking to /usr/local will fail as 3.7 is already there.

**Comment by MikeMcQuaid** [comment]
@imichka Can you elaborate a bit more with step-by-step instructions of what to run and what the output is? This should be fixable in Homebrew/brew (and I'm game for doing that).

**Comment by iMichka** [comment]
Yes, of course. From a clean state:
`brew install python`
`brew install python@3.8`

Then pull this PR.

`brew upgrade -s python@3.8`

You need to build from source as the bottle is not published, but in real life once this PR is merge the bottle will be there and the `-s` is not needed.

This upgrade step will fail because it tries to install `python@3.8` in the path, and will conflict with the already installed Python 3.7 version.

**Comment by SMillerDev** [comment]
@MikeMcQuaid did you ever fix this?

**Comment by MikeMcQuaid** [comment]
No.

**Comment by henryiii** [comment]
Salt needs to be removed from the PR, then (it's 3.8 now), and jupyterlab was already removed (also 3.8). Other than that, is this just waiting on a fix in Hombrew/brew?

**Comment by iMichka** [comment]
Yes; I can rebase my PR to remove salt from here, that should be fast.

The only remaining issue we have is described in my comment above: https://github.com/Homebrew/homebrew-core/pull/54912#issuecomment-639384165

As long as this is not fixed in brew we can't make the transition.

I might also have a look at the code later on when I have more time.

**Comment by MikeMcQuaid** [comment]
> Yes, of course. From a clean state:
> `brew install python`
> `brew install python@3.8`
> 
> Then pull this PR.
> 
> `brew upgrade -s python@3.8`

I did this but:

> This upgrade step will fail because it tries to install `python@3.8` in the path, and will conflict with the already installed Python 3.7 version.

@iMichka This didn't happen to me. Can you paste the full `brew list --versions python python@3.8; brew upgrade --debug --build-from-source python@3.8; brew list --versions python python@3.8` output with the errors you've seen.

**Comment by iMichka** [comment]
```
brew list --versions python python@3.8
python 3.7.7
python@3.8 3.8.3
```


```
brew upgrade --debug --build-from-source python@3.8
...
==> ./configure --prefix=/usr/local/Cellar/python@3.8/3.8.3_1 --enable-ipv6 --datarootdir=/usr/local/Cellar/python@3.8/3.8.3_1/
==> make
==> make install PYTHONAPPSDIR=/usr/local/Cellar/python@3.8/3.8.3_1
==> make frameworkinstallextras PYTHONAPPSDIR=/usr/local/Cellar/python@3.8/3.8.3_1/share/python@3.8
unzip -qq -o /Users/michkapopoff/Library/Caches/Homebrew/downloads/e5199bbf2a8e2450a866445f256b9ca0f890309f843300835d2c4d3ebf9769d3--setuptools-46.0.0.zip -d /private/tmp/d20200702-73897-18dyf6a
cp -pR /private/tmp/d20200702-73897-18dyf6a/setuptools-46.0.0/. /private/tmp/python@3.8--setuptools-20200702-73897-7gvw5k/setuptools-46.0.0
tar xof /Users/michkapopoff/Library/Caches/Homebrew/downloads/81c6f6c92bea8093b41c52486c1d47040fdaa166eba7820a14b14fc6fe4e1fd8--pip-20.0.2.tar.gz -C /private/tmp/d20200702-73897-1ggdeap
cp -pR /private/tmp/d20200702-73897-1ggdeap/pip-20.0.2/. /private/tmp/python@3.8--pip-20200702-73897-18f5umy/pip-20.0.2
tar xof /Users/michkapopoff/Library/Caches/Homebrew/downloads/553d732b4cf117039b60aab1f75ac3e6f6869a9fa3127b4107984fa5b084a00b--wheel-0.34.2.tar.gz -C /private/tmp/d20200702-73897-1oupxt1
cp -pR /private/tmp/d20200702-73897-1oupxt1/wheel-0.34.2/. /private/tmp/python@3.8--wheel-20200702-73897-4uwkbm/wheel-0.34.2
Error: The `brew link` step did not complete successfully
The formula built, but is not symlinked into /usr/local
Could not symlink bin/2to3
Target /usr/local/bin/2to3
is a symlink belonging to python. You can unlink it:
  brew unlink python

To force the link and overwrite all conflicting files:
  brew link --overwrite python@3.8

To list all files that would be deleted:
  brew link --overwrite --dry-run python@3.8

Possible conflicting files are:
/usr/local/bin/2to3 -> /usr/local/Cellar/python/3.7.7/bin/2to3
/usr/local/bin/idle3 -> /usr/local/Cellar/python/3.7.7/bin/idle3
/usr/local/bin/pydoc3 -> /usr/local/Cellar/python/3.7.7/bin/pydoc3
/usr/local/bin/python3 -> /usr/local/Cellar/python/3.7.7/bin/python3
/usr/local/bin/python3-config -> /usr/local/Cellar/python/3.7.7/bin/python3-config
/usr/local/share/man/man1/python3.1 -> /usr/local/Cellar/python/3.7.7/share/man/man1/python3.1
/usr/local/lib/pkgconfig/python3.pc -> /usr/local/Cellar/python/3.7.7/lib/pkgconfig/python3.pc
/usr/local/Frameworks/Python.framework/Headers -> /usr/local/Cellar/python/3.7.7/Frameworks/Python.framework/Headers
/usr/local/Frameworks/Python.framework/Python -> /usr/local/Cellar/python/3.7.7/Frameworks/Python.framework/Python
/usr/local/Frameworks/Python.framework/Resources -> /usr/local/Cellar/python/3.7.7/Frameworks/Python.framework/Resources
/usr/local/Frameworks/Python.framework/Versions/Current -> /usr/local/Cellar/python/3.7.7/Frameworks/Python.framework/Versions/Current
==> Changing dylib ID of /usr/local/Cellar/python@3.8/3.8.3_1/Frameworks/Python.framework/Versions/3.8/Python
  from /usr/local/Cellar/python@3.8/3.8.3_1/Frameworks/Python.framework/Versions/3.8/Python
    to /usr/local/opt/python@3.8/Frameworks/Python.framework/Versions/3.8/Python
==> /usr/local/Cellar/python@3.8/3.8.3_1/bin/python3 -s setup.py --no-user-cfg install --force --verbose --install-scripts=/usr
==> /usr/local/Cellar/python@3.8/3.8.3_1/bin/python3 -s setup.py --no-user-cfg install --force --verbose --install-scripts=/usr
==> /usr/local/Cellar/python@3.8/3.8.3_1/bin/python3 -s setup.py --no-user-cfg install --force --verbose --install-scripts=/usr
/usr/local/Homebrew/Library/Homebrew/brew.rb (Formulary::FormulaLoader): loading /usr/local/Homebrew/Library/Taps/homebrew/homebrew-core/Formula/gettext.rb
==> Caveats
```

```
brew list --versions python python@3.8
python 3.7.7
python@3.8 3.8.3_1
```

**Comment by MikeMcQuaid** [comment]
@iMichka Thanks. I'd suggest using the `link_overwrite` DSL to allow all of those conflicting files to be overwritten. This does not require a Homebrew/brew change.

**Comment by iMichka** [comment]
Nice. Will try it out.

**Comment by iMichka** [comment]
Looks like `link_overwrite` is not working for this case.

It goes into this else statement and returns false, so nothing is done with the conflicting files:
https://github.com/Homebrew/brew/blob/master/Library/Homebrew/formula.rb#L1121

`keg.name` == `python` in that case, and I think the issue is that python.rb has been deleted.

**Comment by Rylan12** [comment]
It looks like `Formulary.factory("python")` follows the `python` alias and points to `python@3.8`, causing `link_overwrite` to return false because it doesn't want to overwrite `python@3.8`'s files.

Would a solution be to allow `link_overwrite` to succeed if the file it's trying to overwrite belongs to the formula currently being installed?

**Comment by MikeMcQuaid** [comment]
> Would a solution be to allow `link_overwrite` to succeed if the file it's trying to overwrite belongs to the formula currently being installed?

I would say in this case it shouldn't even need `link_overwrite` but: yes.

**Comment by iMichka** [comment]
See https://github.com/Homebrew/brew/pull/7915 for a possible fix.

**Comment by iMichka** [comment]
Tested with ly brew fix, `link_overwrite` now works. We just need a new release.

**Comment by Rylan12** [comment]
Just for my own clarification: a release is necessary so that all users will get the latest `brew` changes (including the necessary `link_overwrite` change). Merging this PR without the release would mean that non-dev users who don't have the latest `brew` changes will still get the latest `python@3.8` changes but without the necessary `link_overwrite` changes, causing `python@3.8` to link incorrectly.

**Comment by iMichka** [comment]
Yes exactly. If you have the DEV flag set as homebrew developer, you are on the latest master and live on the bleeding edge. But shipping this PR without a proper released will break all our user's Python's as they stay on fixed release tags.

File: Formula/python@3.7.rb [hunk 3]
```diff
@@ -1,9 +1,8 @@
-class Python < Formula
+class PythonAT37 < Formula
   desc "Interpreted, interactive, object-oriented programming language"
   homepage "https://www.python.org/"
-  url "https://www.python.org/ftp/python/3.7.8/Python-3.7.8.tar.xz"
-  sha256 "43a543404b363f0037f89df8478f19db2dbc0d6f3ffee310bc2997fa71854a63"
-  head "https://github.com/python/cpython.git"
+  url "https://www.python.org/ftp/python/3.7.7/Python-3.7.7.tar.xz"
```
Comment (MikeMcQuaid) [inline Formula/python@3.7.rb:4]: Oops 😭

**Review by MikeMcQuaid — APPROVED** [review]

**Review by MikeMcQuaid — APPROVED** [review]

**Pull request closed without merging**